src/main/scala/mesosphere/marathon/core/group/impl/GroupManagerImpl.scala

```diff
@@ -84,43 +84,54 @@ class GroupManagerImpl(
-    change: (RootGroup) => RootGroup, version: Timestamp, force: Boolean, toKill: Map[PathId, Seq[Instance]]): Future[DeploymentPlan] = {
+    change: (RootGroup) => Either[T, RootGroup],
+    version: Timestamp, force: Boolean, toKill: Map[PathId, Seq[Instance]]): Future[Either[T, DeploymentPlan]] = try {
 
     // All updates to the root go through the work queue.
-    val deployment = serializeUpdates {
-      async {
-        logger.info(s"Upgrade root group version:$version with force:$force")
-
-        val from = rootGroup()
-        val unversioned = assignDynamicServicePorts(from, change(from))
-        val to = GroupVersioningUtil.updateVersionInfoForChangedApps(version, from, unversioned)
-        Validation.validateOrThrow(to)(RootGroup.rootGroupValidator(config.availableFeatures))
-        val plan = DeploymentPlan(from, to, version, toKill)
-        Validation.validateOrThrow(plan)(DeploymentPlan.deploymentPlanValidator())
-        logger.info(s"Computed new deployment plan:\n$plan")
-        await(groupRepository.storeRootVersion(plan.target, plan.createdOrUpdatedApps, plan.createdOrUpdatedPods))
-        await(deploymentService.get().deploy(plan, force))
-        await(groupRepository.storeRoot(plan.target, plan.createdOrUpdatedApps, plan.deletedApps, plan.createdOrUpdatedPods, plan.deletedPods))
-        logger.info(s"Updated groups/apps/pods according to plan ${plan.id}")
-        // finally update the root under the write lock.
-        root := plan.target
-        plan
+    val result = serializeUpdates {
+      logger.info(s"Upgrade root group version:$version with force:$force")
+
+      val from = rootGroup()
+      change(from) match {
+        case Left(left) =>
+          Future.successful(Left(left))
+        case Right(changed) =>
+          async {
+            val unversioned = assignDynamicServicePorts(from, changed)
+            val to = GroupVersioningUtil.updateVersionInfoForChangedApps(version, from, unversioned)
+            Validation.validateOrThrow(to)(RootGroup.rootGroupValidator(config.availableFeatures))
+            val plan = DeploymentPlan(from, to, version, toKill)
+            Validation.validateOrThrow(plan)(DeploymentPlan.deploymentPlanValidator())
+            logger.info(s"Computed new deployment plan:\n$plan")
+            await(groupRepository.storeRootVersion(plan.target, plan.createdOrUpdatedApps, plan.createdOrUpdatedPods))
+            await(deploymentService.get().deploy(plan, force))
+            await(groupRepository.storeRoot(plan.target, plan.createdOrUpdatedApps, plan.deletedApps, plan.createdOrUpdatedPods, plan.deletedPods))
+            logger.info(s"Updated groups/apps/pods according to plan ${plan.id}")
+            // finally update the root under the write lock.
+            root := plan.target
+            Right(plan)
+          }
       }
     }
-    deployment.onComplete {
-      case Success(plan) =>
+
+    result.onComplete {
+      case Success(Right(plan)) =>
```

Comment:
Either is new to me... however it took only a few mins to understand the concept... I can see it's value... but I really dislike it here!   a lot.     It is easy and reasonable that you as the author know what is happening here...but it is easily lost...   The "meaning" of Left or Right needs to be more clearly documented and should be standardized such that Left for 1 use of Either is not Right for another.

Reply:
I agree with Ken here. `Left` is meant to designate a failure and provide some information about the failure. This is its intention in Haskell, from which `Either` and all were borrowed. Also, from the Scala docs: `Convention dictates that Left is used for failure and Right is used for success`

Reply:
Either is a great way to deal with non-panic level failures in a type-safe way (IE, validation failures). The code is legible to me. Is it just an issue of familiarity, and not necessarily illegibility?